### ISSUE TITLE
Support ES6's Iterators

### ISSUE THREAD

**Issue by dhritzkiv** [caolan/async#839]
ES6's Map and Sets feel like they could fit as an input data type, as they can be iterated over easily.
I found myself looking to use async with a `Map` earlier today, but had to convert it to an object and back.

Example with a Map:

```javascript
const numberMap = new Map();
var obj1 = {num: 1};
var obj2 = {num: 2};
numberMap.set(1, obj1);
numberMap.set(2, obj2);

async.map(numberMap, function(item, callback) {
    item.num *= 2;
    callback(null, item);
}, function(err, newNumberMap) {
   console.log(newNumberMap)
   //Map { 1 => { num: 2 }, 2 => { num: 4 } }
   console.log(newNumberMap instanceof Map)
   //true;
});
```

Example with a Set:

```javascript
const numberObjectSet = new Set();
var obj1 = {num: 1};
var obj2 = {num: 2};
numberObjectSet.add(obj1);
numberObjectSet.add(obj2);

async.map(numberObjectSet, function(item, callback) {
    item.num *= 2;
    callback(null, item);
}, function(err, newNumberObjectSet) {
   console.log(newNumberObjectSet);
   //Set { { num: 2 }, { num: 4} }
   console.log(newNumberObjectSet instanceof Set)
   //true;
});
```

Possibly related to #579

**Comment by aearly** [comment]
I like the idea.  You can get decent support today, just by using the spread operator:

```js
async.map([...numberMap], function (item, callback) {...
```

`async.map` always returns an array, though.  We'd have to create special mapping functions for the various collection types, similar to `_.map` vs. `_.mapValues` in lodash, for Arrays vs. Objects.

**Comment by megawac** [comment]
:+1: it'd be relatively easy to support via some `keyIterator` post #847

**Comment by ex1st** [comment]
New PR for generators https://github.com/caolan/async/pull/1074

**Comment by aearly** [comment]
Closed by #1074!
